onnx_trt_backend.cpp

```diff
@@ -135,18 +137,46 @@ class OnnxTensorRTEvent {
   ~OnnxTensorRTEvent() { cudaEventDestroy(event_); }
 
   onnxStatus Signal() {
-    return (cudaEventRecord(event_, stream_) == cudaSuccess)
-               ? ONNXIFI_STATUS_SUCCESS
-               : ONNXIFI_STATUS_INTERNAL_ERROR;
+    std::lock_guard<std::mutex> guard(mutex_);
+    if (fired_) {
+      return ONNXIFI_STATUS_INVALID_STATE;
+    }
+
+    if (cudaEventRecord(event_, stream_) == cudaSuccess) {
+      fired_ = true;
+      return ONNXIFI_STATUS_SUCCESS;
+    } else {
+      return ONNXIFI_STATUS_INTERNAL_ERROR;
+    }
   }
 
   onnxStatus Wait() {
+    std::lock_guard<std::mutex> guard(mutex_);
     return (cudaEventSynchronize(event_) == cudaSuccess)
                ? ONNXIFI_STATUS_SUCCESS
                : ONNXIFI_STATUS_INTERNAL_ERROR;
   }
 
+  onnxStatus CheckState(onnxEventState* state) {
+    std::lock_guard<std::mutex> guard(mutex_);
+    if (!fired_) {
+      *state = ONNXIFI_EVENT_STATE_NONSIGNALLED;
+      return ONNXIFI_STATUS_SUCCESS;
+    }
+
+    auto rt = cudaEventQuery(event_);
+    if (rt == cudaSuccess || rt == cudaErrorNotReady) {
+      *state = ONNXIFI_EVENT_STATE_SIGNALLED;
+      return ONNXIFI_STATUS_SUCCESS;
+    } else {
+      *state = ONNXIFI_EVENT_STATE_INVALID;
+      return ONNXIFI_STATUS_INVALID_STATE;
+    }
+  }
+
 private:
+  std::mutex mutex_;
+  std::atomic<bool> fired_{false};
```

Comment:
Why do you need these? Can't you just query `event_`?

Reply:
`cudaEventQuery` does not tell you whether `cudaEventRecord` has been called or not.